python/lsst/pipe/analysis/utils.py

```diff
@@ -1281,20 +1289,18 @@ def makeBadArray(catalog, flagList=[], onlyReadStars=False, patchInnerOnly=True,
         bad |= ~catalog["detect_isPatchInner"]
     if "detect_isTractInner" in schema and tractInnerOnly:
         bad |= ~catalog["detect_isTractInner"]
+
     if "deblend_scarletFlux" in schema:
-        # Exclude parents that are blends of multiple sources
-        bad |= ((catalog["parent"] == 0) & (catalog["deblend_nChild"] > 1))
-        # Exclude isolated sources not passed to deblender
-        if excludeIsolatedNotModel:
+        if "detect_isDeblendedSource" in schema:
+            if useScarletModelSource:  # use the scarlet model-based results
+                bad |= ~catalog["detect_isDeblendedModelSource"]
+            else:  # use the non-model scarlet results
+                bad |= ~catalog["detect_isDeblendedSource"]
+        else:  # The following is just for backwards compatibility with older scarlet catalogs
+            # Exclude parents that are blends of multiple sources
+            bad |= ((catalog["parent"] == 0) & (catalog["deblend_nChild"] > 1))
+            # Exclude isolated sources not passed to deblender
```

Comment:
Similarly for this block, there is no need to check if scarlet was used or not anymore, that was already done by `pipe.tasks.setPrimary`. Just set `bad=~catalog["detect_isDeblendedSource"]` and all of the non-primary sources (with respect to the deblender) will be labeled "bad".

Reply:
Now it's my turn to be confused. :) You have a parameter `useScarletModelForIsolated`, which is the only time that you check against `isDeblendedModelSource`. So I would think that setting `useScarletModelForIsolated` to `true` when scarlet wasn't run would fail, but that seems like something that should produce an error anyway. But I agree that you need to check if this is a scarlet catalog now, since scarlet omits the isolated sources not passed to the deblender, which is unique to the scarlet dataset.

But I'm curious about your check for the `bad` sources as related to blending. You flagged `good` sources (in the old deblender) as those with `deblend_nChild == 0`, but now you check that `(catalog["parent"] == 0) & (catalog["deblend_nChild"] > 1)` which is just a longer way of writing the same thing, right? Is this to serve as an extra check that they are the same? Otherwise, as I said, you can just use `bad |= ~catalog["detect_isDeblendedSource"]`, unless there's something else that I'm missing.

Reply:
For the former, yeah, `useScarletModelForIsolated` is ignored if using a non-`scarlet`-based catalog.  I added that to the docstring (I think that's sufficient as it's not a config parameter, so it would be "one of us" changing that in the call to `makeBadArray` from within the task).  In doing that, I also noticed that, as is (now was!), such a comment would've also applied to the `excludeSkipped` parameter, but this check should apply for both deblenders, so I moved it outside of the `scarlet`-specific conditioning (it was an oversight that we weren't selecting against them previously...and I'm now wondering if this condition should be added to the `isPrimary` flag (maybe something to consider on DM-16178)?   [Many of them end up effectively excluded because measurement algorithms fail on them...but not _all_ of them...]

For the latter, that is simply there for the case where the scarlet catalog was pre-DM-29087, as identified by the presence of the `deblend_scarletFlux` column but lack of the `detect_isDeblendedSource` column.  So, in all practicality, I left it there so that Eric's run of the `compareCoaddAnalyais` scripts would work for the `w14` (new scarlet flags) vs. `w10` (old scarlet flags).  Indeed, it looks like that ran just fine, as you can see by the `compare-*` plots in, e.g. https://lsst.ncsa.illinois.edu/~emorgan2/w_2021_14_qaplots/HSC-I/tract-9813/
I've added a comment in the code noting this.